### ISSUE TITLE
RSA keyGen results failed to verify

### ISSUE THREAD

**Issue by kwong-cygnacom** [usnistgov/ACVP-Server#108]
**environment**
Demo

**testSessionId**
177760

**vsId**
559135

**Algorithm registration**
[
  {
    "acvVersion": "1.0"
  },
  {
    "isSample": false,
    "publishable": true,
    "algorithms": [
      {
        "algorithm": "RSA",
        "mode": "keyGen",
        "revision": "FIPS186-4",
        "prereqVals": [
          {
            "algorithm": "SHA",
            "valValue": "same"
          },
          {
            "algorithm": "DRBG",
            "valValue": "same"
          }
        ],
        "infoGeneratedByServer": false,
        "pubExpMode": "fixed",
        "fixedPubExp": "010001",
        "keyFormat": "crt",
        "capabilities": [
          {
            "randPQ": "B.3.6",
            "properties": [
              {
                "modulo": 2048,
                "primeTest": [
                  "tblC3"
                ]
              }
            ]
          },
          {
            "randPQ": "B.3.6",
            "properties": [
              {
                "modulo": 3072,
                "primeTest": [
                  "tblC3"
                ]
              }
            ]
          },
          {
            "randPQ": "B.3.6",
            "properties": [
              {
                "modulo": 4096,
                "primeTest": [
                  "tblC3"
                ]
              }
            ]
          }		  
        ]
      }	  
    ]
  }
]

**Endpoint in which the error is experienced**
https://demo.acvts.nist.gov/acvp/v1/testSessions/177760/vectorSets/559135/results

**Expected behavior**
A customer's implementation uses a fixed value of bitlens. The results were expected to pass, but we got an error. Is it the value too small or the same value not allowed for 4 bitlens or something else?

The server returned this error message:
[
  {
    "acvVersion": "1.0"
  },
  {
    "error": "Value _entropyBigIntegers is less than minInclusive"
  }
]

**Additional context**
Uploaded the responses [RSA-keyGen-request-559135-response.json.txt](https://github.com/usnistgov/ACVP-Server/files/6504955/RSA-keyGen-request-559135-response.json.txt)

**Comment by Kritner** [comment]
There's a fix for the next release related to this particular vector set, as we should still be processing the vector set rather than failing out early like we're currently doing.

The issue in this case is because at least one value provided didn't meet the lower bound of a value as required by the modulo.

**Comment by kwong-cygnacom** [comment]
When would the next release be available?

**Comment by Kritner** [comment]
No concrete date for the release, and it would be going to just demo first.  I'd imagine it'll be there sometime next week.  Just to clarify though, the release will only "successfully" process the vector set in question in such a way that it would be able to return the pass/fail per test case.  The issue with some of the values being send being under the lower bound of acceptable values will still need correction in order to get a fully passing validation of the vector set.

FWIW this is the check that was failing within the several test cases I double checked:

![image](https://user-images.githubusercontent.com/8814983/119135825-edfbad00-ba0c-11eb-8cea-09ac54096214.png)

If it would be helpful, I can post a validation file here (or through nfiles) that runs this particular vector set against the corrected code in the works to go out in the release.

**Comment by kwong-cygnacom** [comment]
That will be helpful. You can either post it here or through nfiles.  Thanks.

**Comment by Kritner** [comment]
Note I did not validate if *all* of the failures were due to the bounds check:

```json
{
  "vsId": 559135,
  "disposition": "failed",
  "tests": [
    {
      "tcId": 1,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 2,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 3,
      "result": "passed"
    },
    {
      "tcId": 4,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 5,
      "result": "passed"
    },
    {
      "tcId": 6,
      "result": "passed"
    },
    {
      "tcId": 7,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 8,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 9,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 10,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 11,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 12,
      "result": "passed"
    },
    {
      "tcId": 13,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 14,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 15,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 16,
      "result": "passed"
    },
    {
      "tcId": 17,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 18,
      "result": "passed"
    },
    {
      "tcId": 19,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 20,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 21,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 22,
      "result": "passed"
    },
    {
      "tcId": 23,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 24,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 25,
      "result": "passed"
    },
    {
      "tcId": 26,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 27,
      "result": "passed"
    },
    {
      "tcId": 28,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 29,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 30,
      "result": "passed"
    },
    {
      "tcId": 31,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 32,
      "result": "passed"
    },
    {
      "tcId": 33,
      "result": "passed"
    },
    {
      "tcId": 34,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 35,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 36,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 37,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 38,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 39,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 40,
      "result": "passed"
    },
    {
      "tcId": 41,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 42,
      "result": "passed"
    },
    {
      "tcId": 43,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 44,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 45,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 46,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 47,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 48,
      "result": "passed"
    },
    {
      "tcId": 49,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 50,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 51,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 52,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 53,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 54,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 55,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 56,
      "result": "passed"
    },
    {
      "tcId": 57,
      "result": "passed"
    },
    {
      "tcId": 58,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 59,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 60,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 61,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 62,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 63,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 64,
      "result": "passed"
    },
    {
      "tcId": 65,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 66,
      "result": "passed"
    },
    {
      "tcId": 67,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 68,
      "result": "passed"
    },
    {
      "tcId": 69,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 70,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 71,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 72,
      "result": "failed",
      "reason": "Unable to resolve deferred crypto: Key is not a valid key"
    },
    {
      "tcId": 73,
      "result": "passed"
    },
    {
      "tcId": 74,
      "result": "passed"
    },
    {
      "tcId": 75,
      "result": "passed"
    }
  ]
}
```

**Comment by Kritner** [comment]
This change is on demo in release [v1.1.0.18](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.18)

**Comment by Kritner** [comment]
This change has been released to prod in release [v1.1.0.18](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.18).

**Comment by angelshadow** [comment]
> Note I did not validate if _all_ of the failures were due to the bounds check:
> 
> ```json
> {
>   "vsId": 559135,
>   "disposition": "failed",
>   "tests": [
>     {
>       "tcId": 1,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 2,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 3,
>       "result": "passed"
>     },
>     {
>       "tcId": 4,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 5,
>       "result": "passed"
>     },
>     {
>       "tcId": 6,
>       "result": "passed"
>     },
>     {
>       "tcId": 7,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 8,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 9,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 10,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 11,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 12,
>       "result": "passed"
>     },
>     {
>       "tcId": 13,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 14,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 15,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 16,
>       "result": "passed"
>     },
>     {
>       "tcId": 17,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 18,
>       "result": "passed"
>     },
>     {
>       "tcId": 19,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 20,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 21,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 22,
>       "result": "passed"
>     },
>     {
>       "tcId": 23,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 24,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 25,
>       "result": "passed"
>     },
>     {
>       "tcId": 26,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 27,
>       "result": "passed"
>     },
>     {
>       "tcId": 28,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 29,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 30,
>       "result": "passed"
>     },
>     {
>       "tcId": 31,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 32,
>       "result": "passed"
>     },
>     {
>       "tcId": 33,
>       "result": "passed"
>     },
>     {
>       "tcId": 34,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 35,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 36,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 37,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 38,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 39,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 40,
>       "result": "passed"
>     },
>     {
>       "tcId": 41,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 42,
>       "result": "passed"
>     },
>     {
>       "tcId": 43,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 44,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 45,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 46,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 47,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 48,
>       "result": "passed"
>     },
>     {
>       "tcId": 49,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 50,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 51,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 52,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 53,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 54,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 55,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 56,
>       "result": "passed"
>     },
>     {
>       "tcId": 57,
>       "result": "passed"
>     },
>     {
>       "tcId": 58,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 59,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 60,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 61,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 62,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 63,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 64,
>       "result": "passed"
>     },
>     {
>       "tcId": 65,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 66,
>       "result": "passed"
>     },
>     {
>       "tcId": 67,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 68,
>       "result": "passed"
>     },
>     {
>       "tcId": 69,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 70,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 71,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 72,
>       "result": "failed",
>       "reason": "Unable to resolve deferred crypto: Key is not a valid key"
>     },
>     {
>       "tcId": 73,
>       "result": "passed"
>     },
>     {
>       "tcId": 74,
>       "result": "passed"
>     },
>     {
>       "tcId": 75,
>       "result": "passed"
>     }
>   ]
> }
> ```

I have the same issue that "reason": "Unable to resolve deferred crypto: Key is not a valid key". What does "bounds check" mean? And how should I fix it ?
